src/lock/optiql.cpp

```diff
@@ -181,10 +200,26 @@ OptiQL::LockX()  //
 {
   auto qid = GetQID();
   auto *qnode = &(_qnodes[qid]);
+  const auto new_tail = (static_cast<uint64_t>(qid) << kQIDShift) | kXLock;
 
-  throw std::runtime_error{"not implemented yet"};
+  auto cur = lock_.load(kRelaxed);
+  while (true) {
+    qnode->ver.store(cur & kLockMask, kRelaxed);
+    if (lock_.compare_exchange_weak(cur, new_tail, kAcqRel, kRelaxed)) break;
+    CPP_UTILITY_SPINLOCK_HINT
+  }
+
+  if ((cur & kLockMask) != kNoLocks) {
+    // wait until predecessor gives up the lock
+    auto *pred_qnode = &(_qnodes[(cur & kQIDMask) >> kQIDShift]);
+    pred_qnode->next.store(qnode, kRelease);
+    while (qnode->ver.load(kRelaxed) & kLockMask) {
```

Comment:
フラグをatomicに立てるだけでよくなったので，こちらもフェンスは不要になります．

```suggestion
    pred_qnode->next.store(qnode, kRelaxed);
    while (!qnode->hold_lock.load(kRelaxed)) {
```